File: src/domain/tile_merge.ts
// map_width and map_height are the dimensions of the map
// is_wall_f checks if a tile is a wall

interface  Rect {
  start_x: number;
  start_y: number;
  end_x: number;
  end_y: number;
}

const rectangles: Rect[] = []; // Each rectangle covers a grid of wall tiles

const map_width = 4;
const map_height = 4;

const walls = new Map<number, Set<number>>();

walls.set(3, new Set([0]));

function is_wall_f(x: number, y: number) {
  return !(walls.has(x) && walls.get(x)?.has(y));
}

for (let x = 0; x < map_width; ++x) {
  let start_y = -1;
  let end_y = -1;
  for (let y = 0; y < map_height; ++y) {
    if (is_wall_f(x, y)) {
      if (start_y === -1) {
        start_y = y;
      }
      end_y = y;
    } else if (start_y !== -1) {
      const overlaps = rectangles
        .filter(r => {
          return r.end_x === x - 1 &&
            start_y <= r.start_y &&
            end_y >= r.end_y;
        })
        .sort((r1, r2) => {
          return r1.start_y - r2.start_y;
        });
      overlaps.forEach(r => {
        if (start_y < r.start_y) {
          rectangles.push({
            start_x: x,
            start_y,
            end_x: x,
            end_y: r.start_y - 1
          });
          start_y = r.start_y;
        }

        if (start_y === r.start_y) {
          r.end_x = r.end_x + 1;

          if (end_y === r.end_y) {
            start_y = -1;
            end_y = -1;
          } else if (end_y > r.end_y) {
            start_y = r.end_y + 1;
          }
        }
      });

      if (start_y !== -1) {
        rectangles.push({
          start_x: x,
          start_y,
          end_x: x,
          end_y
        });
        start_y = -1;
        end_y = -1;
      }
    }
  }

  if (start_y !== -1) {
    rectangles.push({
      start_x: x,
      start_y,
      end_x: x,
      end_y
    });
    start_y = -1;
    end_y = -1;
  }
}

rectangles.forEach(r => console.log(r));

File: src/domain/tile_merge.lua
-- map_width and map_height are the dimensions of the map
-- is_wall_f checks if a tile is a wall
 
local rectangles = {} -- Each rectangle covers a grid of wall tiles
 
for x = 0, map_width - 1 do
    local start_y
    local end_y
 
    for y = 0, map_height - 1 do
        if is_wall_f(x, y) then
            if not start_y then
                start_y = y
            end
            end_y = y
        elseif start_y then
            local overlaps = {}
            for _, r in ipairs(rectangles) do
                if (r.end_x == x - 1)
                  and (start_y <= r.start_y)
                  and (end_y >= r.end_y) then
                    table.insert(overlaps, r)
                end
            end
            table.sort(
                overlaps,
                function (a, b)
                    return a.start_y < b.start_y
                end
            )
 
            for _, r in ipairs(overlaps) do
                if start_y < r.start_y then
                    local new_rect = {
                        start_x = x,
                        start_y = start_y,
                        end_x = x,
                        end_y = r.start_y - 1
                    }
                    table.insert(rectangles, new_rect)
                    start_y = r.start_y
                end
 
                if start_y == r.start_y then
                    r.end_x = r.end_x + 1
 
                    if end_y == r.end_y then
                        start_y = nil
                        end_y = nil
                    elseif end_y > r.end_y then
                        start_y = r.end_y + 1
                    end
                end
            end
 
            if start_y then
                local new_rect = {
                    start_x = x,
                    start_y = start_y,
                    end_x = x,
                    end_y = end_y
                }
                table.insert(rectangles, new_rect)
 
                start_y = nil
                end_y = nil
            end
        end
    end
 
    if start_y then
        local new_rect = {
            start_x = x,
            start_y = start_y,
            end_x = x,
            end_y = end_y
        }
        table.insert(rectangles, new_rect)
 
        start_y = nil
        end_y = nil
    end
end